specs/opcode/1dSAR.md

```diff
@@ -0,0 +1,177 @@
+a64s_lo[idx] = a64s[idx] % p_lo
+a64s_hi[idx] = a64s[idx] / p_lo
+```
+
+Under this special case of `shift < 64`, `b64s` could be calculated as:
+```
+b64s[0] = a64s_hi[0] + a64s_lo[1] * p_hi
+b64s[1] = a64s_hi[1] + a64s_lo[2] * p_hi
+b64s[2] = a64s_hi[2] + a64s_lo[3] * p_hi
+b64s[3] = a64s_hi[3] + p_top
+```
+
+#### Common case
+
+For common case, we should add more variables as:
+```
+shf0 = bytes_to_fq(shift.le_bytes[:1])
+shf_lt256 = compare_word(shift, RLC(256))
+shf_div64_eq0 = is_zero(shf_div64)
+shf_div64_eq1 = is_zero(shf_div64 - 1)
+shf_div64_eq2 = is_zero(shf_div64 - 2)
+shf_div64_eq3 = is_zero(shf_div64 - 3)
+```
+
+Then `b64s` could be calculated as:
+
+* Initialization: It should be `[MAX_U64] * 4` if `a` is negative, `[0] * 4` otherwise.
+* `b64s[k]`: It could be calculated by `a64s_hi[k + shf_div64] + a64s_lo[k + shf_div64 + 1] * p_hi` where `k < 3 - shf_div64`.
+* `b64s[3 - shf_div64]`: It could be calculated by `a64s_hi[3] + p_top`.
+
+#### Circuit Constraints
+
+Now putting things together, the constraints can be constructed as follows:
+
+1. `a64s` and `b64s` constraints:
+
+* First calculate `shf_lt256` as `compare_word(shift, RLC(256))`.
+* `a64s[idx]`: It should be equal to `from_bytes(a[8 * idx : 8 * (idx + 1)])` where idx in `(0, 1, 2, 3)`.
+* `b64s[idx] * shf_lt256 + is_neg * (1 - shf_lt256) * MAX_U64`: It should be equal to `from_bytes(b[8 * idx : 8 * (idx + 1)])` where idx in `(0, 1, 2, 3)`.
+
+2. `a64s_lo` and `a64s_hi` constraints:
+
+* `a64s[idx]`: It should be equal to `a64s_lo[idx] + a64s_hi[idx] * p_lo`.
+* `a64s_lo[idx]`: It should always be less than `p_lo` (`a64s_lo[idx] < p_lo`).
+* `a64s_hi[idx]`: It should always be less than `p_hi` (`a64s_hi[idx] < p_hi`).
+
+3. Merge constraints:
+
+* First create four `IsZero` gadgets:
+```
+shf_div64_eq0 = is_zero(shf_div64)
+shf_div64_eq1 = is_zero(shf_div64 - 1)
+shf_div64_eq2 = is_zero(shf_div64 - 2)
+shf_div64_eq3 = is_zero(shf_div64 - 3)
+```
+
+* `b64s[0]` should be equal to:
+```
+(a64s_hi[0] + a64s_lo[1] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq1 +
+  (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq2 +
+  (a64s_hi[3] + p_top) * shf_div64_eq3 +
+  is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2 - shf_div64_eq3)
+```
+
+* `b64s[1]` should be equal to:
+```
+(a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq1 +
+  (a64s_hi[3] + p_top) * shf_div64_eq2 +
+  is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2)
+```
+
+* `b64s[2]` should be equal to:
+```
+(a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[3] + p_top) * shf_div64_eq1 +
+  is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1)
+```
+
+* `b64s[3]` should be equal to:
+```
+(a64s_hi[3] + p_top) * shf_div64_eq0 +
+  is_neg * MAX_U64 * (1 - shf_div64_eq0)
+```
+
+4. `shift[0]` constraint:
+
+* `shift[0]`: It should be equal to `shf_mod64 + shf_div64 * 64`.
```

Comment:
As this https://github.com/privacy-scaling-explorations/zkevm-specs/pull/263#discussion_r1058985665 mentioned, I replaced `shift[0] or shf0` and `shf_lt256`with `shf_lo` and `shf_hi_is_zero`. And they could be calculated as:
```
shf_lo, shf_hi = word_to_lo_hi(shift)
shf_div64 = shf_lo.n // 64
shf_mod64 = shf_lo.n % 64
shf_hi_is_zero = instruction.is_zero(shf_hi)
```
And `b64s` could be constrained as below. If `shf_hi != 0`, then `shift` should be greater than `256` (shift right for more than 256-bit), `b == 2**256 - 1`
 (`0xffff...`) if `is_neg == 1`, `b == 0` otherwise (`is_neg == 0`).
```
instruction.constrain_equal(
    instruction.select(shf_hi_is_zero, b64s[idx], is_neg * MAX_U64),
    instruction.bytes_to_fq(b.le_bytes[offset : offset + N_BYTES_U64]),
)
```

Reply:
And added constraint of `shf_mod64 < 64` as:
```
shf_mod64_lt_64, _ = instruction.compare(shf_mod64, FQ(64), 1)
instruction.constrain_equal(shf_mod64_lt_64, FQ(1))
```
Also updated it to Markdown doc:
```
4. `shift` constraint:

* `shf_mod64`: It should be less than 64.
* `shf_lo`: It should be equal to `shf_mod64 + shf_div64 * 64`.
```